# Regression Model with Artificial Neural Network to Predict value of real estates based on certain Attributes and Comparing the Same with a Multiple Linear Regression Model

In [1]:
import tensorflow as tf
import pandas as pd
import numpy as np

## Attribute Information:

####    1. CRIM :            per capita crime rate by town
####    2. ZN :                proportion of residential land zoned for lots over 25,000 sq.ft.
####    3. INDUS :         proportion of non-retail business acres per town
####    4. CHAS :          Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
####    5. NOX :            nitric oxides concentration (parts per 10 million)
####    6. RM :              average number of rooms per dwelling
####    7. AGE  :           proportion of owner-occupied units built prior to 1940
####    8. DIS   :            weighted distances to five Boston employment centres
####    9. RAD  :           index of accessibility to radial highways
####    10. TAX  :          full-value property-tax rate per  (10,000 dollars)
####    11. PTRATIO :   pupil-teacher ratio by town
####    12. B    :            1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
####    13. LSTAT :       % lower status of the population
####    14. MEDV   :      Median value of owner-occupied homes in $1000's 

In [2]:
dataset = pd.read_csv('data.csv')
dataset

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

# Data Preprocessing 

In [3]:
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

## Converting y into 2d array from 1d array 

In [4]:
y =y.reshape(len(y), 1)

## Taking Care of Missing Values

In [5]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values = np.nan,strategy='mean')
imputer.fit(X[:, 5:6])
imputer.fit(y)
X[:, 5:6] = imputer.transform(X[:, 5:6])
y=imputer.transform(y)

# Splitting the dataset into Training and Test Set 

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Scaling the features for better performance 

In [7]:
from sklearn.preprocessing import StandardScaler
sc_x_1 = StandardScaler()
sc_y = StandardScaler()
sc_x_2 = StandardScaler()
X_train[:, 0:3]= sc_x_1.fit_transform(X_train[:,0:3])
X_train[:, 4:13]= sc_x_2.fit_transform(X_train[:,4:13])
y_train = sc_y.fit_transform(y_train)

# Creating Artificial Neural Network Model

In [8]:
ann = tf.keras.models.Sequential()

## Input Layer 

In [9]:
ann.add(tf.keras.layers.Dense(500, input_shape=(13,), activation='relu'))

## Hidden Layers 

In [10]:
ann.add(tf.keras.layers.Dense(500, activation='relu'))

In [11]:
ann.add(tf.keras.layers.Dense(500, activation='relu'))
ann.add(tf.keras.layers.Dense(500, activation='relu'))
ann.add(tf.keras.layers.Dense(500, activation='relu'))
ann.add(tf.keras.layers.Dense(500, activation='relu'))
ann.add(tf.keras.layers.Dense(500, activation='relu'))

In [12]:
ann.add(tf.keras.layers.Dense(500, activation='relu'))

In [13]:
ann.add(tf.keras.layers.Dense(500, activation='relu'))

In [14]:
ann.add(tf.keras.layers.Dense(500, activation='relu'))

In [15]:
ann.add(tf.keras.layers.Dense(500, activation='relu'))

In [16]:
ann.add(tf.keras.layers.Dense(500, activation='relu'))

In [17]:
ann.add(tf.keras.layers.Dense(500, activation='relu'))

In [18]:
ann.add(tf.keras.layers.Dense(500, activation='relu'))

## Output Layer 

In [19]:
ann.add(tf.keras.layers.Dense(1, activation='linear'))

## Compiling all the input, hidden and output layers 

In [20]:
ann.compile(optimizer='Adam', loss='mean_squared_error', metrics=['accuracy'])

## Training the data with Artificial Neural Network 

In [21]:
ann.fit(X_train, y_train, epochs=1000, verbose=1)

Train on 404 samples
Epoch 1/1000
404/404 [==============================] - 2s 6ms/sample - loss: 0.9817 - accuracy: 0.0000e+00
Epoch 2/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.8207 - accuracy: 0.0000e+00
Epoch 3/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.4166 - accuracy: 0.0000e+00
Epoch 4/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.4098 - accuracy: 0.0000e+00
Epoch 5/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.2326 - accuracy: 0.0000e+00
Epoch 6/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.2442 - accuracy: 0.0000e+00
Epoch 7/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.1635 - accuracy: 0.0000e+00
Epoch 8/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.1226 - accuracy: 0.0000e+00
Epoch 9/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0837 - accuracy: 0.0000e+00
Epoch 1

404/404 [==============================] - 1s 2ms/sample - loss: 0.0126 - accuracy: 0.0000e+00
Epoch 76/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0218 - accuracy: 0.0000e+00
Epoch 77/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0138 - accuracy: 0.0000e+00
Epoch 78/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0117 - accuracy: 0.0000e+00
Epoch 79/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0149 - accuracy: 0.0000e+00
Epoch 80/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0157 - accuracy: 0.0000e+00
Epoch 81/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0121 - accuracy: 0.0000e+00
Epoch 82/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0072 - accuracy: 0.0000e+00
Epoch 83/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0082 - accuracy: 0.0000e+00
Epoch 84/1000
404/404 [==========

404/404 [==============================] - 1s 2ms/sample - loss: 0.0496 - accuracy: 0.0000e+00
Epoch 149/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0371 - accuracy: 0.0000e+00
Epoch 150/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0446 - accuracy: 0.0000e+00
Epoch 151/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0315 - accuracy: 0.0000e+00
Epoch 152/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0169 - accuracy: 0.0000e+00
Epoch 153/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0097 - accuracy: 0.0000e+00
Epoch 154/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0150 - accuracy: 0.0000e+00
Epoch 155/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0089 - accuracy: 0.0000e+00
Epoch 156/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0069 - accuracy: 0.0000e+00
Epoch 157/1000
404/404 [=

404/404 [==============================] - 1s 2ms/sample - loss: 0.0104 - accuracy: 0.0000e+00
Epoch 223/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0116 - accuracy: 0.0000e+00
Epoch 224/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0135 - accuracy: 0.0000e+00
Epoch 225/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0216 - accuracy: 0.0000e+00
Epoch 226/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0719 - accuracy: 0.0000e+00
Epoch 227/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0888 - accuracy: 0.0000e+00
Epoch 228/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0668 - accuracy: 0.0000e+00
Epoch 229/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0414 - accuracy: 0.0000e+00
Epoch 230/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0362 - accuracy: 0.0000e+00
Epoch 231/1000
404/404 [=

404/404 [==============================] - 1s 2ms/sample - loss: 0.0023 - accuracy: 0.0000e+00
Epoch 297/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0012 - accuracy: 0.0000e+00
Epoch 298/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0011 - accuracy: 0.0000e+00
Epoch 299/1000
404/404 [==============================] - 1s 2ms/sample - loss: 8.7977e-04 - accuracy: 0.0000e+00
Epoch 300/1000
404/404 [==============================] - 1s 2ms/sample - loss: 8.2973e-04 - accuracy: 0.0000e+00
Epoch 301/1000
404/404 [==============================] - 1s 2ms/sample - loss: 8.7663e-04 - accuracy: 0.0000e+00
Epoch 302/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0013 - accuracy: 0.0000e+00
Epoch 303/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0017 - accuracy: 0.0000e+00
Epoch 304/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0015 - accuracy: 0.0000e+00
Epoch 305/100

404/404 [==============================] - 1s 2ms/sample - loss: 5.2663e-04 - accuracy: 0.0000e+00
Epoch 371/1000
404/404 [==============================] - 1s 2ms/sample - loss: 6.0285e-04 - accuracy: 0.0000e+00
Epoch 372/1000
404/404 [==============================] - 1s 2ms/sample - loss: 7.1798e-04 - accuracy: 0.0000e+00
Epoch 373/1000
404/404 [==============================] - 1s 2ms/sample - loss: 6.2128e-04 - accuracy: 0.0000e+00
Epoch 374/1000
404/404 [==============================] - 1s 2ms/sample - loss: 6.3314e-04 - accuracy: 0.0000e+00
Epoch 375/1000
404/404 [==============================] - 1s 2ms/sample - loss: 5.7150e-04 - accuracy: 0.0000e+00
Epoch 376/1000
404/404 [==============================] - 1s 2ms/sample - loss: 6.0056e-04 - accuracy: 0.0000e+00
Epoch 377/1000
404/404 [==============================] - 1s 2ms/sample - loss: 5.2765e-04 - accuracy: 0.0000e+00
Epoch 378/1000
404/404 [==============================] - 1s 2ms/sample - loss: 5.3656e-04 - accuracy: 

404/404 [==============================] - 1s 2ms/sample - loss: 8.3216e-04 - accuracy: 0.0000e+00
Epoch 444/1000
404/404 [==============================] - 1s 2ms/sample - loss: 5.8010e-04 - accuracy: 0.0000e+00
Epoch 445/1000
404/404 [==============================] - 1s 2ms/sample - loss: 6.2693e-04 - accuracy: 0.0000e+00
Epoch 446/1000
404/404 [==============================] - 1s 2ms/sample - loss: 6.5360e-04 - accuracy: 0.0000e+00
Epoch 447/1000
404/404 [==============================] - 1s 2ms/sample - loss: 5.9296e-04 - accuracy: 0.0000e+00
Epoch 448/1000
404/404 [==============================] - 1s 2ms/sample - loss: 4.4673e-04 - accuracy: 0.0000e+00
Epoch 449/1000
404/404 [==============================] - 1s 2ms/sample - loss: 3.9700e-04 - accuracy: 0.0000e+00
Epoch 450/1000
404/404 [==============================] - 1s 2ms/sample - loss: 4.6421e-04 - accuracy: 0.0000e+00
Epoch 451/1000
404/404 [==============================] - 1s 2ms/sample - loss: 6.6383e-04 - accuracy: 

404/404 [==============================] - 1s 2ms/sample - loss: 3.4563e-04 - accuracy: 0.0000e+00
Epoch 515/1000
404/404 [==============================] - 1s 2ms/sample - loss: 3.2290e-04 - accuracy: 0.0000e+000s - loss: 3.4174e-04 - ac
Epoch 516/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.3595e-04 - accuracy: 0.0000e+00
Epoch 517/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.9307e-04 - accuracy: 0.0000e+00
Epoch 518/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.2942e-04 - accuracy: 0.0000e+00
Epoch 519/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.9339e-04 - accuracy: 0.0000e+00
Epoch 520/1000
404/404 [==============================] - 1s 2ms/sample - loss: 5.7279e-04 - accuracy: 0.0000e+00
Epoch 521/1000
404/404 [==============================] - 1s 2ms/sample - loss: 5.8111e-04 - accuracy: 0.0000e+000s - loss: 5.9188e-04 - accuracy: 
Epoch 522/1000
404/404 [===================

404/404 [==============================] - 1s 2ms/sample - loss: 3.2155e-04 - accuracy: 0.0000e+00
Epoch 587/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.0337e-04 - accuracy: 0.0000e+00
Epoch 588/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.2423e-04 - accuracy: 0.0000e+00
Epoch 589/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.1267e-04 - accuracy: 0.0000e+00
Epoch 590/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.5463e-04 - accuracy: 0.0000e+00
Epoch 591/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.5738e-04 - accuracy: 0.0000e+00
Epoch 592/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.0023e-04 - accuracy: 0.0000e+000s - loss: 1.0354e-04 - accuracy: 0.0000e+
Epoch 593/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.0780e-04 - accuracy: 0.0000e+00
Epoch 594/1000
404/404 [==============================] - 1s 

404/404 [==============================] - 1s 2ms/sample - loss: 0.0026 - accuracy: 0.0000e+00
Epoch 659/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0018 - accuracy: 0.0000e+00
Epoch 660/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0014 - accuracy: 0.0000e+00
Epoch 661/1000
404/404 [==============================] - 1s 2ms/sample - loss: 9.6546e-04 - accuracy: 0.0000e+00
Epoch 662/1000
404/404 [==============================] - 1s 2ms/sample - loss: 8.8599e-04 - accuracy: 0.0000e+00
Epoch 663/1000
404/404 [==============================] - 1s 2ms/sample - loss: 7.0204e-04 - accuracy: 0.0000e+00
Epoch 664/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0014 - accuracy: 0.0000e+00
Epoch 665/1000
404/404 [==============================] - 1s 2ms/sample - loss: 8.8277e-04 - accuracy: 0.0000e+00
Epoch 666/1000
404/404 [==============================] - 1s 2ms/sample - loss: 5.8864e-04 - accuracy: 0.0000e+00
Epoch

404/404 [==============================] - 1s 2ms/sample - loss: 1.4826e-04 - accuracy: 0.0000e+000s - loss: 1.7090e-04 - 
Epoch 732/1000
404/404 [==============================] - 1s 2ms/sample - loss: 8.5431e-05 - accuracy: 0.0000e+00
Epoch 733/1000
404/404 [==============================] - 1s 2ms/sample - loss: 9.7585e-05 - accuracy: 0.0000e+00
Epoch 734/1000
404/404 [==============================] - 1s 2ms/sample - loss: 8.0265e-05 - accuracy: 0.0000e+00
Epoch 735/1000
404/404 [==============================] - 1s 2ms/sample - loss: 7.6329e-05 - accuracy: 0.0000e+00
Epoch 736/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.3811e-04 - accuracy: 0.0000e+00
Epoch 737/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.1282e-04 - accuracy: 0.0000e+00
Epoch 738/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.9661e-04 - accuracy: 0.0000e+00
Epoch 739/1000
404/404 [==============================] - 1s 2ms/sample - loss:

404/404 [==============================] - 1s 2ms/sample - loss: 0.0010 - accuracy: 0.0000e+00
Epoch 804/1000
404/404 [==============================] - 1s 2ms/sample - loss: 6.4813e-04 - accuracy: 0.0000e+00
Epoch 805/1000
404/404 [==============================] - 1s 2ms/sample - loss: 5.8477e-04 - accuracy: 0.0000e+00
Epoch 806/1000
404/404 [==============================] - 1s 2ms/sample - loss: 4.8839e-04 - accuracy: 0.0000e+00
Epoch 807/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.6811e-04 - accuracy: 0.0000e+00
Epoch 808/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.6684e-04 - accuracy: 0.0000e+00
Epoch 809/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.0116e-04 - accuracy: 0.0000e+00
Epoch 810/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.2297e-04 - accuracy: 0.0000e+00
Epoch 811/1000
404/404 [==============================] - 1s 2ms/sample - loss: 3.0900e-04 - accuracy: 0.00

404/404 [==============================] - 1s 2ms/sample - loss: 2.3882e-04 - accuracy: 0.0000e+00
Epoch 876/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.3233e-04 - accuracy: 0.0000e+00
Epoch 877/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.9924e-04 - accuracy: 0.0000e+00
Epoch 878/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.6782e-04 - accuracy: 0.0000e+00
Epoch 879/1000
404/404 [==============================] - 1s 2ms/sample - loss: 1.5091e-04 - accuracy: 0.0000e+00
Epoch 880/1000
404/404 [==============================] - 1s 2ms/sample - loss: 2.6264e-04 - accuracy: 0.0000e+00
Epoch 881/1000
404/404 [==============================] - 1s 2ms/sample - loss: 4.8686e-04 - accuracy: 0.0000e+00
Epoch 882/1000
404/404 [==============================] - 1s 2ms/sample - loss: 4.3734e-04 - accuracy: 0.0000e+00
Epoch 883/1000
404/404 [==============================] - 1s 2ms/sample - loss: 3.2575e-04 - accuracy: 

404/404 [==============================] - 1s 2ms/sample - loss: 6.8823e-04 - accuracy: 0.0000e+00
Epoch 947/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0012 - accuracy: 0.0000e+00
Epoch 948/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0020 - accuracy: 0.0000e+00
Epoch 949/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0015 - accuracy: 0.0000e+00
Epoch 950/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0013 - accuracy: 0.0000e+00
Epoch 951/1000
404/404 [==============================] - 1s 2ms/sample - loss: 7.4674e-04 - accuracy: 0.0000e+00
Epoch 952/1000
404/404 [==============================] - 1s 2ms/sample - loss: 6.9315e-04 - accuracy: 0.0000e+00
Epoch 953/1000
404/404 [==============================] - 1s 2ms/sample - loss: 6.4271e-04 - accuracy: 0.0000e+00
Epoch 954/1000
404/404 [==============================] - 1s 2ms/sample - loss: 0.0012 - accuracy: 0.0000e+00
Epoch 955

## Testing the ANN on test set

In [22]:
X_test[:, 0:3]= sc_x_1.transform(X_test[:,0:3])
X_test[:, 4:13]= sc_x_2.transform(X_test[:,4:13])

In [23]:
y_pred = sc_y.inverse_transform(ann.predict(X_test))

## Evaluating the ANN model with R-square method

In [24]:
from sklearn.metrics import r2_score
r2_score(y_test,y_pred)

0.7111054102558014

# Creating Multiple Linear Regression Model 

## Training the data with Multiple Linear Regression

In [27]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

LinearRegression()

## Predicting and Testing on Test Set

In [28]:
y_pred_lin = sc_y.inverse_transform(regressor.predict(X_test))

## Evaluating the Multiple Linear Regression Model with R-square Method

In [29]:
r2_score(y_test,y_pred_lin)

0.546464779516093

# For Artificial Neural Network Model we see that the model evaluation with R-square method gives us an accuracy of approximately 71%



# And for Multiple Linear Regression Model the model evaluation with R-square method gives us an accuracy of approximately 54.64%

# Therefore we can conclude that the ANN model gives us greater advantage of Linear Regression Model in case Regression, as Neural Networks nearly mimics the Human Brain